File: edict/managed/agent_logging_test.go
// Copyright 2026 JetBrains s.r.o. Licensed under the Apache License, Version 2.0.

package managed

import (
	"bytes"
	"errors"
	"io"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
)

func TestAgentLoggerAttributesFullOutputAndRedactsRevokedTokens(t *testing.T) {
	store, manager, _ := testStore(t)
	var output, short bytes.Buffer
	logger := NewAgentLogger(store, &output, &short)
	now := time.Now()
	logger.Record(AgentMessage{Time: now, AgentID: "manager-thread", IsRoot: true, Phase: "commentary", Text: "Starting managed analysis\n" + strings.Repeat("Multiline commentary 世界. ", 80)})
	logger.Record(AgentMessage{Time: now, AgentID: "worker-thread", AgentPath: "/root/batch", Phase: "commentary", Text: "Inspecting commit"})
	require.NoError(t, logger.Flush(false))
	require.Contains(t, output.String(), "[edict_manager/-] commentary: Starting managed analysis")
	require.NotContains(t, output.String(), "Inspecting commit")

	grant, err := delegateTestTask(store, manager, store.Plan().Tasks[0].ID, nil, nil)
	require.NoError(t, err)
	_, err = store.StartTask(grant.Token, "/root/batch", grant.Skill) // Some runtimes return paths, others UUIDs.
	require.NoError(t, err)
	require.NoError(t, logger.Flush(false))
	prefix := "[edict-batch-signal-analysis/" + grant.TaskID[:8] + "] "
	require.Contains(t, output.String(), prefix+"commentary: Inspecting commit")

	_, err = store.FinishTask(grant.Token, "completed", "Done")
	require.NoError(t, err)
	hash := strings.Repeat("f", 64)
	long := strings.Repeat("Detailed output. ", 80)
	logger.Record(AgentMessage{Time: now, AgentID: "worker-thread", AgentPath: "/root/batch", Phase: "final_answer",
		Text: long + "\nPlan " + store.Plan().ID + "; hash " + hash + "; tokens " + manager + " " + grant.Token})
	logger.Record(AgentMessage{Time: now, AgentID: "lost-worker", Text: "Could not start"})
	require.NoError(t, logger.Flush(true))
	unwrapped := strings.ReplaceAll(output.String(), "\n    ", "")
	require.Contains(t, unwrapped, prefix+"final: "+long)
	require.Contains(t, unwrapped, "Plan "+store.Plan().ID+"; hash "+hash+"; tokens [redacted] [redacted]")
	require.Contains(t, output.String(), "[unassigned/-] message: Could not start")
	require.NotContains(t, output.String(), manager)
	require.NotContains(t, output.String(), grant.Token)
	require.NotContains(t, output.String(), "agent=")
	require.Equal(t, 1, strings.Count(output.String(), "Inspecting commit"))
	previous := output.String()
	require.NoError(t, logger.Flush(true))
	require.Equal(t, previous, output.String(), "flushing must not duplicate messages")
	require.Equal(t, output.String(), short.String(), "commentary and final messages must be identical in both logs")
}

type failingAgentLogWriter struct{}

func (failingAgentLogWriter) Write([]byte) (int, error) { return 0, errors.New("disk full") }

func TestAgentLoggerReportsWriteFailure(t *testing.T) {
	store, _, _ := testStore(t)
	for _, shortFails := range []bool{false, true} {
		var output bytes.Buffer
		var full, short io.Writer = failingAgentLogWriter{}, &output
		if shortFails {
			full, short = short, full
		}
		logger := NewAgentLogger(store, full, short)
		logger.Record(AgentMessage{Time: time.Now(), AgentID: "root", IsRoot: true, Text: "Output"})
		require.ErrorContains(t, logger.Flush(true), "disk full")
		previous := output.String()
		require.ErrorContains(t, logger.Flush(true), "disk full")
		require.Equal(t, previous, output.String(), "retrying after a write failure must not duplicate messages")
	}
}

File: edict/kotlin/src/test/kotlin/org/jetbrains/qodana/edict/logging/AgentLoggingTest.kt
package org.jetbrains.qodana.edict.logging

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption.*
import java.time.Instant
import kotlin.test.*
import kotlinx.serialization.json.*
import org.jetbrains.qodana.edict.common.flag
import org.jetbrains.qodana.edict.common.json
import org.jetbrains.qodana.edict.common.obj
import org.jetbrains.qodana.edict.common.text
import org.jetbrains.qodana.edict.common.wireJson
import org.jetbrains.qodana.edict.mcp.McpServer
import org.jetbrains.qodana.edict.model.Step
import org.jetbrains.qodana.edict.runtime.CodexAgentCollector
import org.jetbrains.qodana.edict.runtime.CodexRunner
import org.jetbrains.qodana.edict.store.Store
import org.jetbrains.qodana.edict.support.batch
import org.jetbrains.qodana.edict.support.launch
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir

class AgentLoggingTest {
    @TempDir lateinit var directory: Path

    @Test fun `collector attributes nested output defers early workers and redacts revoked tokens`() {
        Store(directory.resolve("state")).use { store ->
            val created = store.createPlan("Extract", listOf(Step("edict-batch-signal-analysis", "Commit")))
            val batch = store.launch(created.token, created.plan.tasks.single().id, "edict-batch-signal-analysis", emptyList(), emptyList())
            val child = store.addTask(batch.token, "edict-signal-analysis", "Inspect")
            val leaf = store.delegate(batch.token, child.id, emptyList(), emptyList(), "\$managed-edict-signal-analysis\nRead /skills/managed-edict-signal-analysis/SKILL.md")
            val home = directory.resolve("home")
            val sessions = home.resolve("sessions")
            Files.createDirectories(sessions)
            val stdout = directory.resolve("stdout.jsonl")
            Files.writeString(stdout, """{"type":"thread.started","thread_id":"root-thread"}""" + "\n")
            val logs = directory.resolve("log")
            val collector = CodexAgentCollector(home, stdout, AgentLogger(store, logs))
            fun append(file: String, value: JsonObject) = Files.writeString(sessions.resolve(file), wireJson.encodeToString(JsonObject.serializer(), value) + "\n", CREATE, APPEND)
            fun meta(file: String, id: String, parent: String = "", path: String = "") = append(file, buildJsonObject {
                put("type", "session_meta")
                putJsonObject("payload") {
                    put("id", id)
                    if (parent.isEmpty()) put("source", "exec") else putJsonObject("source") {
                        putJsonObject("subagent") { putJsonObject("thread_spawn") { put("parent_thread_id", parent); put("agent_path", path) } }
                    }
                }
            })
            fun message(text: String, role: String = "assistant", phase: String = "commentary") = buildJsonObject {
                put("type", "response_item"); put("timestamp", Instant.now().toString())
                putJsonObject("payload") {
                    put("type", "message"); put("role", role); put("phase", phase)
                    putJsonArray("content") { add(buildJsonObject { put("type", "output_text"); put("text", text) }) }
                }
            }
            meta("root.jsonl", "root-thread")
            meta("batch.jsonl", store.plan()!!.tasks.first().agentId, "root-thread")
            meta("leaf.jsonl", "leaf-thread", store.plan()!!.tasks.first().agentId, "/root/batch/leaf")
            meta("unrelated.jsonl", "other-thread")
            append("root.jsonl", message("Manager commentary"))
            append("batch.jsonl", message("Batch commentary"))
            append("leaf.jsonl", message("Early leaf output"))
            append("unrelated.jsonl", message("Unrelated output"))
            append("root.jsonl", message("Private prompt", role = "user"))
            append("root.jsonl", message("Private reasoning", phase = "analysis"))
            collector.scan()
            assertFalse(Files.readString(logs.resolve("edict-agents.log")).contains("Early leaf output"))
            store.readTask(leaf.token)
            store.startTask(leaf.token, "/root/batch/leaf", leaf.skill)
            collector.scan()
            store.finishTask(leaf.token, "completed", "Done")
            val digest = "f".repeat(64)
            append("leaf.jsonl", message("Leaf finished $digest ${leaf.token} ${created.token}\n" + "Unicode 世界😀 ".repeat(50), phase = "final_answer"))
            val final = wireJson.encodeToString(JsonObject.serializer(), message("Manager final", phase = "final_answer"))
            Files.writeString(sessions.resolve("root.jsonl"), final.take(final.length / 2), APPEND)
            collector.scan()
            assertFalse(Files.readString(logs.resolve("edict-agents.log")).contains("Manager final"))
            Files.writeString(sessions.resolve("root.jsonl"), final.drop(final.length / 2), APPEND)
            collector.scan(final = true)
            collector.scan(final = true)
            val full = Files.readString(logs.resolve("edict-agents.log"))
            val unwrapped = full.replace("\n    ", "")
            assertContains(unwrapped, "[edict-signal-analysis/${leaf.taskId.take(8)}] final: Leaf finished $digest [REDACTED] [REDACTED]")
            listOf("Manager commentary", "Batch commentary", "Early leaf output", "Manager final").forEach {
                assertEquals(1, Regex(Regex.escape(it)).findAll(full).count())
            }
            listOf("Unrelated output", "Private prompt", "Private reasoning", leaf.token, created.token).forEach { assertFalse(full.contains(it)) }
            full.lineSequence().forEach { assertTrue(it.codePointCount(0, it.length) <= 120) }
            assertEquals(full, Files.readString(logs.resolve("edict-agent-short.log")))
        }
    }

    @Test fun `MCP logs contain complete assignments and summaries without capability leaks`() {
        Store(directory.resolve("state")).use { store ->
            val logs = directory.resolve("log")
            val server = McpServer(store, logs = logs)
            fun call(name: String, arguments: JsonObject): JsonObject {
                val result = server.call(name, arguments)
                assertEquals(false, result.flag("isError"))
                return result.obj("structuredContent")
            }
            val created = call("edict_plan_create", buildJsonObject {
                put("request", "Extract"); put("steps", json.encodeToJsonElement(listOf(Step("edict-batch-signal-analysis", "Commit"))))
            })
            val task = store.plan()!!.tasks.single()
            val prompt = "\$managed-edict-batch-signal-analysis\nRead /skills/managed-edict-batch-signal-analysis/SKILL.md and inspect HEAD."
            val delegated = call("edict_delegate", buildJsonObject { put("token", created.text("token")); put("taskId", task.id); put("prompt", prompt) })
            val token = delegated.text("token")
            call("edict_task_get", buildJsonObject { put("token", token) })
            call("edict_task_start", buildJsonObject { put("token", token); put("agentId", "worker"); put("skill", task.skill) })
            call("edict_task_finish", buildJsonObject { put("token", token); put("status", "completed"); put("result", "Done") })
            val full = Files.readString(logs.resolve("edict-agents.log"))
            val short = Files.readString(logs.resolve("edict-agent-short.log"))
            assertContains(full.replace("\n    ", ""), prompt.replace("\n", ""))
            assertContains(full, "[${task.skill}/${task.id.take(8)}] mcp: edict_task_finish response:")
            assertFalse(short.contains("response:"))
            assertContains(short, "Started task")
            listOf(full, short).forEach { assertFalse(it.contains(token)); assertFalse(it.contains(created.text("token"))) }
        }
    }

    @Test fun `runtime failure still flushes agent messages`() {
        Store(directory.resolve("state")).use { store ->
            val executable = directory.resolve("codex-stub")
            Files.writeString(executable, """
                #!/bin/sh
                mkdir -p "${'$'}CODEX_HOME/sessions"
                printf '%s\n' '{"type":"thread.started","thread_id":"failed-root"}'
                printf '%s\n' '{"type":"session_meta","payload":{"id":"failed-root","source":"exec"}}' '{"timestamp":"2026-09-23T16:00:00Z","type":"response_item","payload":{"type":"message","role":"assistant","phase":"commentary","content":[{"type":"output_text","text":"Partial output before failure"}]}}' > "${'$'}CODEX_HOME/sessions/root.jsonl"
                exit 7
            """.trimIndent())
            executable.toFile().setExecutable(true)
            val logs = directory.resolve("log")
            val runtime = CodexRunner(directory, directory, store.root, "http://127.0.0.1/mcp", executable.toString(), "test", AgentLogger(store, logs))
            runtime.prepare()
            assertContains(assertFailsWith<IllegalStateException> { runtime.run("test") }.message!!, "Codex exited 7")
            assertContains(Files.readString(logs.resolve("edict-agents.log")), "[edict_manager/-] commentary: Partial output before failure")
        }
    }
}
